Window manager changes

HaTaX2 opened the pull request:
A window manager class has been added and configuration entries to the general config and compute device configs to allow initial placement of windows.  Multiple monitors are supported as well.

No changes made to the GUI to make setting these values, you'll need to edit the general config and benchmark config JSON files.

Master pull request notes for maintainers:
These changes allow a user to have mining worker windows always start up in a particular location on a specified monitor.  By default the settings are all -1 which disables this functionality to prevent unexpected behavior.  Initial mining windows also set to not steal foreground focus regardless of these settings. If the debug console is active, it's snapped to the bottom of the main window when it opens. For users that use the PC while mining, these settings will keep windows where they belong and keep them from stealing focus while they're working in other windows.  Modifications to the GUI can be made after these changes are determined to function well and without interference.

HaTaX2 committed:
Window manager changes

A window manager class has been added and configuration entries to the general config and compute device configs to allow initial placement of windows.  Multiple monitors are supported as well.

No changes made to the GUI to make setting these values, you'll need to edit the general config and benchmark config JSON files.

drkskwlkr commented:
@HaTaX2 I find tremendous benefits in this functionality and am interested to test it, but I don't think I am capable of building a custom branch from source on my own. Are you inclined to provide a x64 binary based on latest NHML build that I could run for some time & test for UX/bugs?

HaTaX2 committed:
small fix to move window only once

Use the form's startposition setting to determine if an attempt should be made to move the window.  InitMainConfigGUIData() is called more than once at startup and if the debug window was enabled it would move the main window instead after initial startup.

HaTaX2 commented:
If you'd like to test it, go for it!  This link will work for 2 days if you'd like to download the compiled Release build: https://expirebox.com/download/605cd92df5c735d6f09894b1215fc4fc.html

Just as a note, when you run it the first time it will add the config keys into General.json and benchmark_*.json files.  Close out NHML before changing the config values from -1 to whatever you like, then when you start it back up those values will take effect.

HaTaX2 committed:
Update README.md

Configuration entries used by window_manager have been added with descriptions.

drkskwlkr commented:
@HaTaX2 Got it! I am off to work now but will test in the evening and during the weekend. Thank you!

drkskwlkr commented:
Hey @HaTaX2 — the modification works beautifully! I can finally park the NHML window where I want it.

I am curious, any chance you could look into providing support for virtual desktops as well? This would be an even better solution for a single-monitor system: you can have your NHML window on your main desktop, and the miners/debug windows parked just a `Ctrl+WinKey+Arrow` away.

I have no idea how difficult would that be but I see .NET does have an API for this: [https://www.codeproject.com/Articles/7666/Desktop-Switching](https://www.codeproject.com/Articles/7666/Desktop-Switching)

I did ask a similar question to @DillonN last summer (see #51) and he said he'd look into it, but I have no idea whether the VD idea proved not to be viable for execution or it was simply superseded by other more urgent matters.

HaTaX2 commented:
I started looking into the virtual desktop functionality but it will take a bit more work to integrate that.  I think it's doable for sure, but it will take a bit of work to add it.  I can definitely see the benefit of adding this functionality however, right now NHML runs happily on a virtual desktop, but the miner windows are spawned on whatever desktop is currently active. 

When I get a chance to put something together for this I'll update here.

drkskwlkr commented:
@DillonN I've been running the modified version by @HaTaX2 for 10 days now & can vouch that it works as expected and introduces no glitches or side effects. It offers a huge convenience by making it easier to pin the NHML windows to fixed locations.

Will you consider merging this PR into the next pre-release version of the software so that more people can test it?

The author is willing to add changes to GUI to make this feature easier to set up but IMHO such effort is not even necessary as adjusting window position is not something that needs to be done only once per desktop, and editing a config file is fine.

HaTaX2 commented:
@drkskwlkr Thanks for the testing and confirmation that things work as expected!  If I did the GUI work I'd probably want to go overboard and allow you to change settings and 'test' them by having it place the windows (or a mock/fake miner window) to help with positioning.  As it is, you'll probably end up restarting NHML a few times to get it exactly where you want but once it's set you really don't need to mess with it again.

I've had a few friends testing this and I've been running it on a few different boxes and haven't run into any issues thus far. I'm still working on a few other things for NHML (IE: adding support to save previous benchmarks for review); hopefully DillonN & other maintainers will find these changes useful / beneficial.

e239 commented:
plz anyone help me how to download miner
